File: 324.wiggle-sort-ii.py
#
# @lc app=leetcode id=324 lang=python3
#
# [324] Wiggle Sort II
#

# Time:  O(nlogn)
# Space: O(n)

from typing import List


# @lc code=start
class Solution:
    def wiggleSort(self, nums: List[int]) -> None:
        nums.sort()
        mid = (len(nums) - 1) // 2
        nums[::2], nums[1::2] = nums[mid::-1], nums[:mid:-1]
        return

        tmp = sorted(nums)
        left, right = (len(nums) - 1) // 2, len(nums) - 1
        for i in range(len(nums)):
            if i % 2 == 0:
                nums[i] = tmp[left]
                left -= 1
            else:
                nums[i] = tmp[right]
                right -= 1


# @lc code=end

File: 324.wiggle-sort-ii.js
/*
 * @lc app=leetcode id=324 lang=javascript
 *
 * [324] Wiggle Sort II
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @return {void} Do not return anything, modify nums in-place instead.
 */
var wiggleSort = function (nums) {
  const sorted = nums.slice();
  sorted.sort((a, b) => a - b);
  let left = Math.trunc((nums.length - 1) / 2);
  let right = nums.length - 1;
  for (let i = 0; i < nums.length; ++i) {
    nums[i] = i % 2 === 0 ? sorted[left--] : sorted[right--];
  }
};
// @lc code=end
